README
         _______         ___________          ___
       __\___   \_  ____/     /\   /______ ___\_/__            /\
      /    \|     \/  _ \    /--\_____    \\   /  /\          /  \
    _/      |     /    \     \      |/     /      \ \ _ _____/    \_______
    \       |_____\____/     /      /      \____  /\/                    /
     \______|      \  \_____/\______       /\   \/  \   /\NSILOVE 1.09  /____ _
      \     |_______\__\___ \ \    \      /  \___\__/____
       \____|        /____/\_\/\__ /     / __/___ \     /\    ____
                     \    \ \     /     /_/ __   \/    /__\__/  _/ _____ ___
                 /\   \____\/    /     //   \    /    /    /    _>/    //__/\
                /  \           _/      \   //   /    /    /     \/    / \__\/
_ _________    /    \_______ _ \___    /_______/    /   _/ \____      \  /
           \  /                 \ \___/\       \_______/\\  \  \______/\/
            \/                   \_\  \ \_______\      \ \\/ \__\     \ \
                                    \__\/        \______\/ h7/dS!\_____\/

# Description

Ansilove is a set of tools to convert ANSi and artscene-related file formats
into PNG images, supporting ANSi (.ANS), PCBOARD (.PCB), BiNARY (.BIN), ADF
(.ADF), iDF (.IDF), TUNDRA (.TND) and XBiN (.XB) formats.

It includes an all purposes command line converter, along with a library to
use on web pages for doing "on the fly" conversions. It creates size optimized
4-bit PNG files and supports SAUCE (Standard Architecture for Universal Comment
Extentions), 80x25 and 80x50 PC fonts (including all the 14 MS-DOS charsets
and some unofficial ones), Amiga fonts, and iCE colors.

It is written in PHP, and can thus be run on any supported system : the
command line converter can be used under Linux, Mac OS X, Windows, and so
far and so on.

Ansilove doesn’t require any Web server : you can use the bundled command
line converter to convert files locally on your computer.

# Features

- ANSi (.ANS) format support
- PCBOARD (.PCB) format support
- BiNARY (.BIN) format support
- ADF (.ADF) format support (Artworx)
- iDF (.IDF) format support (iCE Draw)
- TUNDRA (.TND) format support (http://tundradraw.sourceforge.net)
- XBiN (.XB) format support (http://www.acid.org/info/xbin/xbin.htm)
- Small output file size (4-bit PNG)
- SAUCE (Standard Architecture for Universal Comment Extentions)
- 80x25 font support
- 80x50 font support
- Amiga font support
- iCE colors support

## Supported charsets

  IBM PC (Code page 437)
  Baltic (Code page 775)
  Cyrillic (Code page 855)
  French Canadian (Code page 863)
  Greek (Code pages 737 and 869)
  Hebrew (Code page 862)
  Icelandic (Code page 861)
  Latin-1 (Code page 850)
  Latin-2 (Code page 852)
  Nordic (Code page 865)
  Portuguese (Code page 860)
  Russian (Code page 866)
  Turkish (Code page 857)
  Armenian
  Persian (Iran System encoding standard)

# License

Ansilove is released under a MIT-style license. See LICENSE file for details.

# CONTACT

Project Homepage : http://ansilove.sourceforge.net

Sister project : http://www.ascii-codes.com


ChangeLog
                _ __ ____ ___________              __________ ____ __ _
                  .________         //_ ANSilOVE _\\        ._____
        ____/\ ___|      _/___/\   __    _____.   _ ____/\ _|     \ ____
       /  .__\\   \      |  _ //___ _)_. \_   |____      \\ |    // ___/
      /   __/  \  \      |__ \/    \   |_ |   |    \ |     \     / __)____
     /____|     \_|\_____|_________/__ `/ |______ //_______/____/______  /
          |______/    _ __ __________ \/ _______ \/ ________ __ _  yop \/


Ansilove/PHP 1.12 (2017-04-19)

- Adding a new example ANSi (Thanks Otium)
- Adding a new example ASCii (Thanks Konami)
- Adding a new example ANSi (Thanks Andy Herbert)
- Adding a new example ANSi (Thanks TCF)
- Adding a new example XBin (Thanks Enzo)
- Relicensed under the BSD 2-Clause license
- Use env to find PHP
- Switch examples pages to HTML5
- Add deprecation notices



Ansilove/PHP 1.11 (2014-03-26)

- Fixed a bug in the TUNDRA renderer
- Fixed inconsistency when using Amiga fonts: bold characters are now
  displayed only in CED and WorkBench modes
- Removing PHP warning about date.timezone settings when logging errors
- Removing PHP warning displayed when input file doesn't exist
- Added 3 example ASCiis (Thanks to: Plur, Spear, Knocturnal, Sk!n)
- Added 9 example ANSis (Thanks to: Knocturnal, Grymmjack, Noches,
  Avenging Angel, Enzo, BYM, Filth, Rad Man)



Ansilove/PHP 1.10 (2012-10-08)

- Added new and updated documentation file in AsciiDoc format (Also available
  in HTML and PDF)
- Using proper bitshift instead of pow for ADF / IDF / XBIN font processing
- Skipping 'Set mode' and 'Reset mode' sequences to correctly render files
  saved with recent versions of PabloDraw



Ansilove/PHP 1.09 (2011-11-22)

- Added functions in the ANSi loader to split the output into several PNG files
  (SPLIT_* options in the config file)
- Added support for Amiga ANSi (Bold, Italics, and Underlined characters)
- Memory usage optimization (Some data structures modified, support for 65535
  lines of text)
- Relicensed under a MIT-style license
- Added Terminus font support (http://terminus-font.sourceforge.net)



Ansilove/PHP 1.08 (2011-07-12)

- Added new improved Amiga fonts (With corrected aspect ratio and 100% accurate
  characters)
- Reordered font palettes to follow the ANSi color order rather than the EGA one
- Added an error message logger (ANSILOVE_LOG_FILE option in the config file,
  default: ansilove.log)
- Added an option to continue rendering input files when the substitude (SUB)
  character is encountered (SUBSTITUTE_BREAK option in the config file,
  default: 1 => Stop when encountering SUB character)
- Added an option to allow enabling/disabling 80th column wrapping
  (WRAP_COLUMN_80 option in the config file, default: 1 => Wrap)
- Changed default background color value for CED mode (170,170,170) instead
  of (168,168,168)
- Added a transparent rendering mode to produce output files with transparent
  background (only available in ANSi loader)
- Added Workbench mode to render the input file using Amiga Workbench colors
  (only available in ANSi loader)
- Added scalable thumbnails support (THUMBNAILS_SIZE option in the config file)
- Added an example file for thumbnails rendering (examples/thumbnails.html)
- Added 3 example ASCiis (Thanks to: Cubon, Spidy and Spot)



Ansilove/PHP 1.07 (2009-06-16)

- Added support for thumbnails rendering, with configurable maximum height
- Added customizable output color definitions in the configuration file for
  CED mode rendering
- Modified the ANSi parser to allow ansi sequences with inverted graphic
  rendition values to render correctly



Ansilove/PHP 1.06 (2009-05-21)

- Dumped all the fonts again to make the font collection more homogeneous and
  avoid minor character differences between fonts as much as possible
- Added 6 new PC Fonts (Charsets: Armenian, French Canadian, Greek (CP869),
  Icelandic, Persian, Portuguese)
- Fixed minor color palette problems causing the full intensity value to be
  252 instead of 255
- Added customizable file extensions definitions in the configuration file
  for DIZ mode rendering
- Added filters removing white spaces and empty lines at the beginning and
  end of DIZ files



Ansilove/PHP 1.05 (2009-04-29)

- Added support for PCBoard (.PCB) format
- Updated the documentation (README file), to cover options related to the
  PCBoard loader (PCB codes stripping)
- Fixed a bug preventing the CED rendering mode to operate correctly. Ansi and
  PCBoard files can now be rendered in CED mode again
- Added 1 example ANSi and 1 example ASCii



Ansilove/PHP 1.04 (2009-04-07)

- Added 9 new PC Fonts (Charsets: Baltic, Cyrillic, Greek, Hebrew, Latin1,
  Latin2, Nordic, Russian, Turkish)
- Fixed .DIZ file extension detection in the online converter, uploaded .DIZ
  files are now rendered with the exact needed number of columns
- Fixed some uninitialized variables warnings
- Added 1 example ASCii
- Added a loader (ansilove.bat) for the command line converter under Windows
- Converted the HTML templates to XHTML (ansilove.html and Ansilove online)
- Modified the Ansilove Online upload interface and corrected the maximum
  length value for the columns argument
- Updated the documentation (README file), which is now providing more detailed
  examples about all the possible options, and informations about command line
  converter usage under Windows



Ansilove/PHP 1.03 (2006-02-11)

- Added input sanitization in the loaders (load_*.php) to prevent
  disclosure of unauthorized content by converting files accessible
  by the webserver
- Added a .htaccess to restrict access (Deny to all by default) to the
  'upload' directory (online converter)
- Adding '.ansilove' at the end of the uploaded filenames to prevent code
  injection if the upload directory access isn't restricted, and delete
  the uploaded file after the convertion (online converter)

ansilove.cfg.php
<?PHP
/*****************************************************************************/
/*   ____   __ __   ___. __  __   _______._______                            */
/* __\   \./  \| \. \  |_| \|  |_/    \  |  /  _/                            */
/* \___\__|__|\___|_____/___|____/____/\___/\____/plur                       */
/*                                                                           */
/* Ansilove/PHP 1.12                                                         */
/* Copyright (c) 2003-2017, Frederic Cambus                                  */
/* https://www.ansilove.org                                                  */
/*                                                                           */
/* Created:      2003-07-17                                                  */
/* Last Updated: 2017-02-06                                                  */
/*                                                                           */
/* Ansilove is released under the BSD 2-Clause license.                      */
/* See LICENSE file for details.                                             */
/*                                                                           */
/*****************************************************************************/

/*****************************************************************************/
/* SECURITY WARNING!           SECURITY WARNING!           SECURITY WARNING! */
/*                                                                           */
/* Don't unset the ANSILOVE_FILES_DIRECTORY defined  constant, else it'll be */
/* possible to convert files  laying in the same directory than the loaders, */
/* which could lead to possible security leaks.                              */
/*                                                                           */
/* SECURITY WARNING!           SECURITY WARNING!           SECURITY WARNING! */
/*****************************************************************************/

DEFINE (ANSILOVE_FILES_DIRECTORY,"ansis/");

DEFINE (ANSILOVE_LOG_FILE,"ansilove.log");

DEFINE (PCBOARD_STRIP_CODES,"@POFF@,@WAIT@");

DEFINE (DIZ_EXTENSIONS,".diz,.ion");

DEFINE (SUBSTITUTE_BREAK,"1");
DEFINE (WRAP_COLUMN_80,"1");

DEFINE (CED_BACKGROUND_COLOR,"170,170,170");
DEFINE (CED_FOREGROUND_COLOR,"0,0,0");

DEFINE (WORKBENCH_COLOR_0,"170,170,170");
DEFINE (WORKBENCH_COLOR_1,"0,0,255");
DEFINE (WORKBENCH_COLOR_2,"255,255,255");
DEFINE (WORKBENCH_COLOR_3,"0,255,255");
DEFINE (WORKBENCH_COLOR_4,"0,0,0");
DEFINE (WORKBENCH_COLOR_5,"255,0,255");
DEFINE (WORKBENCH_COLOR_6,"102,136,187");
DEFINE (WORKBENCH_COLOR_7,"255,255,255");

DEFINE (THUMBNAILS_SIZE,"1");
DEFINE (THUMBNAILS_HEIGHT,"0");
DEFINE (THUMBNAILS_TAG,"-thumbnail");

DEFINE (SPLIT,"0");
DEFINE (SPLIT_HEIGHT,"4096");
DEFINE (SPLIT_SEPARATOR,".");
?>
